gradio/chat_interface.py

```diff
@@ -461,18 +461,36 @@ def _generate_chat_title(self, conversation: list[MessageDict]) -> str:
             title = title[:40] + "..."
         return title or "Conversation"
 
+    def make_serializable(self, obj: Any) -> dict | list | str:
+        if isinstance(obj, list):
+            return [self.make_serializable(item) for item in obj]
+        elif isinstance(obj, dict):
+            return {k: self.make_serializable(v) for k, v in obj.items()}
+        elif hasattr(obj, "get_config"):
+            try:
+                if hasattr(obj, "value"):
+                    return obj.value
+            except Exception:
+                return str(obj)
+        return obj
+
     def _save_conversation(
         self,
         index: int | None,
         conversation: list[MessageDict],
         saved_conversations: list[list[MessageDict]],
     ):
         if self.save_history:
+            serializable_conversation = self.make_serializable(conversation)
             if index is not None:
-                saved_conversations[index] = conversation
+                saved_conversations[index] = cast(
```

Comment:
This casting is not ideal as it'll mask any typing issues, would be better if we ensured that make_serializable returned a list[MessageDict].

Reply:
To fix this, changed the `make_serializable` method to return `any`. Since it's recursive and has to to return other the other types (messagedict, component) as well

Reply:
Lmk if this is better, than `cast`

Reply:
This is ok but I think the better approach would be for `make_serializable` to have the correct return type, but inside `make_serializable`, you have an inner function that is called recursively. That way, you still get the guarantees of type checking.